src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
Is it worth the complexity to have one file per file kind? What benefit does it bring? A device with 0:0 returning ENODEV on any operation looked good enough for me.

> That said, our current device cgroup policy for nspawn doesn't permit device node creation for 0:0, we should open this up at the same time.

Good point... In the case of rkt, we cannot have a version of systemd in the container different than the version of systemd-nspawn. So changing the device cgroup policy at the same time should be good enough for our use case.